Migrate list view E2E tests to Playwright

WunderBart opened the pull request:
## What?
Migrate List View E2E tests to Playwright.

## Why?
- Done as a part of the Playwright migration project: https://github.com/WordPress/gutenberg/issues/38851 (priority list),
- Should eventually resolve all [related flaky reports](https://github.com/WordPress/gutenberg/issues?q=is%3Aopen+is%3Aissue+label%3A%22%5BType%5D+Flaky+Test%22+%22specs%2Feditor%2Fvarious%2Flist-view.test.js%22).

## How?
- Used accessible selectors where possible,
- Ensured the UI has caught up when navigating with keyboard,
- Reworded some test titles for consistency,
- Added some comments for clarity.

## Testing Instructions

The output of the following should be all ✅: 

```
npm run test:e2e:playwright -- list-view.spec.js
```

WunderBart committed:
Migrate block list drag&drop test

WunderBart committed:
Migrate first block removal regression test

WunderBart committed:
Don't use the generated snapshot

WunderBart committed:
Migrate previous block selection test

WunderBart committed:
Migrate the next block selection test

WunderBart committed:
Migrate test for removing all blocks

WunderBart committed:
Migrate nested items test

alexstine commented:
@WunderBart #45135 will likely get merged in soon, will this hurt your efforts?

WunderBart committed:
Avoid extra nesting
Don't do it if it's unnecessary for the test to pass as it increases the fragility by bringing another selector into the equation.

WunderBart committed:
Migrate the Home/End keynavigation test

WunderBart committed:
Make test titles consistent

WunderBart committed:
Remove old List View tests (complete migration)

WunderBart committed:
Update a few more comments

WunderBart committed:
Fix the dragging test
Dragged the wrong thing by mistake :)

WunderBart committed:
Move the most repeated part to a helper fn

kevin940726 commented on test/e2e/specs/editor/various/list-view.spec.js:
```diff
@@ -0,0 +1,345 @@
+ */
+const { test, expect } = require( '@wordpress/e2e-test-utils-playwright' );
+
+const openListView = async ( { page, pressKeyWithModifier } ) => {
+	// Open List View with keyboard.
+	await pressKeyWithModifier( 'access', 'o' );
+	// Wait until the current item is selected to ensure UI is ready.
+	await page
+		.getByRole( 'gridcell', {
+			name: /link/i,
+			selected: true,
+		} )
+		.waitFor();
+};
```
It's recommended to create a POM-style util at the bottom of the file, as shown in other tests (`post-editor-template-mode.spec.js` for instance).

kevin940726 replied:
In addition, The second `waitFor` call is usually not needed in Playwright. Actions will auto-wait for the UI to be ready. I think we can just inline this with a single `await pressKeyWithModifier( 'access', 'o' );` call.

WunderBart replied:
> The second waitFor call is usually not needed in Playwright. Actions will auto-wait for the UI to be ready.

I agree it's usually the case, although this `waitFor` is needed in case the `openListView` is immediately followed by another `keypress` (like [here](https://github.com/WordPress/gutenberg/blob/e91096e4c177dae74bf55eb8f699bc0b819ca508/test/e2e/specs/editor/various/list-view.spec.js#L79)), which can kick in before the List View is opened. Does that make sense?

kevin940726 replied:
Oh yeah, that makes sense! I'd still inline it wherever it's needed though, but that's just my opinion and yours works too!

WunderBart replied:
Ah, I actually misunderstood what you meant by inlining - I thought you meant to drop the `waitFor` entirely. Yeah, I was actually on the fence about creating that helper. I think I like inlined version better too, so I'll just revert that commit where I created it! 😄

WunderBart replied:
Updated in https://github.com/WordPress/gutenberg/pull/47919/commits/ff29e8ac82c522a2d1a0401adf22b6caf03f05bb.

kevin940726 commented on test/e2e/specs/editor/various/list-view.spec.js:
```diff
@@ -0,0 +1,345 @@
+			] );
+
+		// Drag the paragraph above the heading.
+		const paragraphBlockItem = page.getByRole( 'gridcell', {
```
Nit: It's more readable to get from the list view than from the whole page.

```js
const listView = page.getByRole( 'region', { name: 'List View' } );
const paragraphBlockItem = listView.getByRole( 'gridcell', { name: 'Paragraph link' } );
```

WunderBart replied:
I actually did that, but then had second thoughts and [removed it](https://github.com/WordPress/gutenberg/pull/47919/commits/23d9daff7e74f9a5fcd40613bbe1a937d8b4c95f). 😅 The reason behind removing it was because it adds an unnecessary selector which can potentially make the test more fragile. I agree, though, that it would increase the readability of the code - I'll bring it back.

kevin940726 replied:
> it adds an unnecessary selector which can potentially make the test more fragile.

I don't think it would make the test more fragile though. On the contrary, I think it will help us catch some unintentional bugs. For instance, imagine there is a gridcell role with the same name that exists in some other part of the page, and the original one for some reason disappears. It's very unlikely though 😅.

WunderBart replied:
I thought about it and assumed the accessible name of the parent (`Block navigation structure`) would change sooner than another `gridcell` with a matching name would appear in the DOM. 😄 That difference is likely negligible, though, so increasing the readability is a good tradeoff. 👍

WunderBart replied:
Updated in https://github.com/WordPress/gutenberg/pull/47919/commits/80a52315cee116a768951862197552aa3809f1c5.

kevin940726 commented on test/e2e/specs/editor/various/list-view.spec.js:
```diff
@@ -0,0 +1,345 @@
+				.getByRole( 'gridcell', {
+					name: 'Image link',
+				} )
+				.locator( 'a' )
```
Could this be rewritten into `getByRole`?

```suggestion
				.getByRole( 'link', { name: 'Image' } )
```

WunderBart replied:
I remember this didn't work for some reason, but I can't recall the reason now. Will doublecheck.

WunderBart replied:
Updated in https://github.com/WordPress/gutenberg/pull/47919#:~:text=Use%20getByRole(%20%27link%27%20)%20instead%20of%20locator(%20%27a%27%20).

WunderBart committed:
Remove obsolete snapshot

Mamaduka commented on test/e2e/specs/editor/various/list-view.spec.js:
```diff
@@ -0,0 +1,345 @@
+				name: 'Image link',
+				selected: true,
+			} )
+			.waitFor();
```
Why do we need these `.waitFor()`s? The locators auto-wait and retry, if I remember correctly.

Mamaduka replied:
I think we can replace these with `await expect(locator).toBeVisible();` this checks the attached/visible DOM element. I find these more explicit. What do you think?

WunderBart replied:
https://github.com/WordPress/gutenberg/pull/47919/commits/2193b792b6ea60076682b8b34a292f7783181c4c

Mamaduka commented on test/e2e/specs/editor/various/list-view.spec.js:
```diff
@@ -0,0 +1,345 @@
+
+		// Start listeninig to console errors.
+		let hasThrownConsoleError = false;
+		page.on( 'console', ( msg ) => {
+			if ( msg.type() === 'error' ) {
+				hasThrownConsoleError = true;
+			}
+		} );
```
Suggestion (non-blocking): We can also use `page.on( 'pagerror' )` to catch just errors - https://playwright.dev/docs/api/class-page#page-event-page-error.

kevin940726 replied:
`pageerror` is probably not the same as console errors, but I didn't try it 😅 .

WunderBart replied:
Will check the `pageerror` and update if it applies here! 👍

WunderBart replied:
Removed the console error check entirely in https://github.com/WordPress/gutenberg/pull/47919/commits/5aa48826c3d979b2d1abc0e17e55ab5650fd94b3. The reason for that is that it would never get to that assertion anyway because the preceding one would catch if an error occurred.

kevin940726 commented on test/e2e/specs/editor/various/list-view.spec.js:
```diff
@@ -0,0 +1,345 @@
+
+		// Ensure console didn't throw an error as reported in
+		// https://github.com/WordPress/gutenberg/issues/38763.
+		await expect( hasThrownConsoleError ).toBeFalsy();
```
I don't think `toBeFalsy()` returns a promise?

```suggestion
		expect( hasThrownConsoleError ).toBeFalsy();
```

WunderBart replied:
Updated in https://github.com/WordPress/gutenberg/pull/47919/commits/5aa48826c3d979b2d1abc0e17e55ab5650fd94b3 (also, see https://github.com/WordPress/gutenberg/pull/47919#discussion_r1103790059)

kevin940726 commented on test/e2e/specs/editor/various/list-view.spec.js:
```diff
@@ -0,0 +1,345 @@
+
+		// List View should have two rows.
+		await expect(
+			page.getByRole( 'gridcell', { name: /link/i } )
```
Nit: I think we can just do string literal here:

```suggestion
			page.getByRole( 'gridcell', { name: 'link' } )
```

Same for other places below.

WunderBart replied:
I think we're not yet at Playwright 1.28, so this would be an [exact match](https://playwright.dev/docs/next/api/class-framelocator#frame-locator-get-by-role-option-exact)? This regex is intended because those `gridcell` names contain the block name as well, e.g. `Image link`, `Header link`, etc. and I want to locate them all.

kevin940726 replied:
I think it would still by default match substring, or it will be a breaking change. I didn't test it though, maybe I'm wrong 😅. Feel free to ignore :)

WunderBart replied:
It does the case-insensitive match but doesn't substring. I've just double-checked and got this:

```
locator.waitFor: Timeout 10000ms exceeded.
=========================== logs ===========================
waiting for selector "role=gridcell[selected=true][name="link"i]" to be visible
============================================================
```

kevin940726 commented on test/e2e/specs/editor/various/list-view.spec.js:
```diff
@@ -0,0 +1,345 @@
+				name: 'Cover link',
+				expanded: true,
+			} )
+			.waitFor();
```
Ditto as what @Mamaduka suggested in https://github.com/WordPress/gutenberg/pull/47919/files#r1102633437.

WunderBart replied:
Updated in https://github.com/WordPress/gutenberg/pull/47919/commits/2193b792b6ea60076682b8b34a292f7783181c4c.

kevin940726 commented on test/e2e/specs/editor/various/list-view.spec.js:
```diff
@@ -0,0 +1,345 @@
+		// Collapse the Cover block.
+		await page
+			.getByRole( 'gridcell', { name: 'Cover link' } )
+			.locator( '.block-editor-list-view__expander[aria-hidden="true"]' )
```
Non-blocking: This is where I think `data-testid` will be more explicit in the source code. Right now we don't seem to have any other options though.

WunderBart replied:
Cool, I wasn't aware we're using the `testid` approach as well. Should I add it to the expander element?

kevin940726 replied:
That will be nice! ❤️ 

TBH, I'm not sure if everyone agrees that we should be using `testid` in this project though. I guess the best way to find out is to start using it 😆 .

WunderBart replied:
Agreed! It feels like a good way to go since it allows us to use the `getBy*` format locator. Updated in https://github.com/WordPress/gutenberg/pull/47919/commits/1891141deb51286996eb5127e795a24a4a790320.

kevin940726 commented on test/e2e/specs/editor/various/list-view.spec.js:
```diff
@@ -0,0 +1,345 @@
+				name: 'Paragraph link',
+				selected: true,
+			} )
+			.waitFor();
```
Let's do `.toBeVisible()` or other explicit assertions here. It's not ideal compared to something like `.toBeSelected()` but Playwright doesn't seem to support it at the time.

WunderBart replied:
Updated in https://github.com/WordPress/gutenberg/pull/47919/commits/2193b792b6ea60076682b8b34a292f7783181c4c.

kevin940726 reviewed (commented):
Awesome work! Thanks a lot!

WunderBart commented:
> @WunderBart #45135 will likely get merged in soon, will this hurt your efforts?

Looks like it will, @alexstine! Thanks for the heads up, I'll keep an eye on that PR. 👍

WunderBart committed:
Revert "Move the most repeated part to a helper fn"

This reverts commit e91096e4c177dae74bf55eb8f699bc0b819ca508.

WunderBart committed:
Revert "Avoid extra nesting"

This reverts commit 23d9daff7e74f9a5fcd40613bbe1a937d8b4c95f.

WunderBart committed:
Replace waitFor with expect.toBeVisible

WunderBart committed:
Add some more locator nesting for readability

WunderBart committed:
Use editor.canvas where appropriate

WunderBart committed:
Remove unnecessary page error check
If the error is actually thrown, the preceding assertion that checks the number of List View items will catch it.

WunderBart committed:
Use getByRole( 'link' ) instead of locator( 'a' )

WunderBart committed:
Use testid for the expander element

WunderBart commented:
@kevin940726 @Mamaduka, it's ready for another go! 🙌

kevin940726 reviewed (approved):
Looks perfect! Thanks a lot! 💯

WunderBart commented:
@alexstine, this one's ready to be shipped. Let me know if I should wait for https://github.com/WordPress/gutenberg/pull/45135 – I don't want to block you, in case you're ready as well.

alexstine commented:
@WunderBart If you can give me one more day, that would be great. I will try to land it soon. Waiting on one final technical review.

Mamaduka reviewed (approved):
Thank you, @WunderBart!

WunderBart committed:
Migrate test added in #45135

WunderBart commented:
@alexstine, I've [migrated](https://github.com/WordPress/gutenberg/pull/47919/commits/4dd325f28becd046b70b9d97e173bd5ca2af7725) the test you've added in #45135. Let me know if it looks good to you!

alexstine commented:
@WunderBart I am sure it is fine. The code diffs are a bit too much for me to check in this PR. I imagine these tests will likely change in the future as the list view itself needs some adjustments after discovery of a couple random focus bugs.

Thanks.

alexstine commented:
@WunderBart Just wanted to see if you are merging this one soon? I am working on more list view improvements which could use E2E tests.
